url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/news/utility/Stemmer.java

```diff
@@ -0,0 +1,562 @@
+package darwInvest.news.utility;
+
+/*
+ * 
+   Porter stemmer in Java. The original paper is in
+
+       Porter, 1980, An algorithm for suffix stripping, Program, Vol. 14,
+       no. 3, pp 130-137,
+
+   See also http://www.tartarus.org/~martin/PorterStemmer
+
+   History:
+
+   Release 1
+
+   Bug 1 (reported by Gonzalo Parra 16/10/99) fixed as marked below.
+   The words 'aed', 'eed', 'oed' leave k at 'a' for step 3, and b[k-1]
+   is then out outside the bounds of b.
+
+   Release 2
+
+   Similarly,
+
+   Bug 2 (reported by Steve Dyrdahl 22/2/00) fixed as marked below.
+   'ion' by itself leaves j = -1 in the test for 'ion' in step 5, and
+   b[j] is then outside the bounds of b.
+
+   Release 3
+
+   Considerably revised 4/9/00 in the light of many helpful suggestions
+   from Brian Goetz of Quiotix Corporation (brian@quiotix.com).
+
+   Release 4
+
+ */
+
+/**
+ * Stemmer, implementing the Porter Stemming Algorithm
+ *
+ * The Stemmer class transforms a word into its root form.  The input
+ * word can be provided a character at time (by calling add()), or at once
+ * by calling one of the various stem(something) methods.
+ * 
+ * @author Martin Porter
+ * @modified Kevin Dolan - for stylistic consistency with darwInvest
+ */
+public class Stemmer
+{  
+	private char[] b;
+	private int i,     /* offset into b */
+	i_end, /* offset to end of stemmed word */
+	j, k;
+	private static final int INC = 50;
+
+	/* unit of size whereby b is increased */
+	public Stemmer()
+	{  
+		b = new char[INC];
+		i = 0;
+		i_end = 0;
+	}
+
+	/**
+	 * Add a character to the word being stemmed.  When you are finished
+	 * adding characters, you can call stem(void) to stem the word.
+	 */
+	public void add(char ch)
+	{  
+		if (i == b.length) {  
+			char[] new_b = new char[i+INC];
+			for (int c = 0; c < i; c++) new_b[c] = b[c];
+			b = new_b;
+		}
+		b[i++] = ch;
+	}
+	
+	
+	/** Adds wLen characters to the word being stemmed contained in a portion
+	 * of a char[] array. This is like repeated calls of add(char ch), but
+	 * faster.
+	 */
+	public void add(char[] w, int wLen)
+	{  
+		if (i+wLen >= b.length) {  
+			char[] new_b = new char[i+wLen+INC];
+			for (int c = 0; c < i; c++) 
+				new_b[c] = b[c];
+			b = new_b;
+		}
+		for (int c = 0; c < wLen; c++) 
+			b[i++] = w[c];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;w&#39; (index: 0 | type: char[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;wLen&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).